File: frontend/types/course.ts
export type CourseVisibility = "private" | "public_template"

export type CourseStatus = "active" | "archived"

export interface Course {
  id: string
  owner_id: string
  title: string
  course_code: string | null
  description: string | null
  subject: string | null
  cover_url: string | null
  visibility: CourseVisibility | string
  status: CourseStatus | string
  created_at: string
  updated_at: string
}

export interface CourseCreatePayload {
  title: string
  course_code?: string
  subject?: string
  description?: string
  visibility: "private"
}

File: backend/app/schemas/course.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


CourseVisibility = Literal["private", "public_template"]
CourseStatusFilter = Literal["active", "archived", "all"]


class CourseCreate(BaseModel):
    title: str = Field(min_length=1, max_length=128)
    course_code: str | None = Field(default=None, max_length=64)
    description: str | None = None
    subject: str | None = Field(default=None, max_length=128)
    cover_url: str | None = None
    visibility: CourseVisibility = "private"


class CourseUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=128)
    course_code: str | None = Field(default=None, max_length=64)
    description: str | None = None
    subject: str | None = Field(default=None, max_length=128)
    cover_url: str | None = None
    visibility: CourseVisibility | None = None


class CourseRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    owner_id: UUID
    title: str
    course_code: str | None = None
    description: str | None = None
    subject: str | None = None
    cover_url: str | None = None
    visibility: str
    status: str
    created_at: datetime
    updated_at: datetime
